text/3556-re-export-symbols.md

```diff
@@ -0,0 +1,175 @@
+
+```
+warning: `#[no_mangle]` will not re-export private items from `extern` block
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature [works today](https://github.com/aidanhs/rust-re-export-lib/) (at least on Linux), showing the following key output which demonstrates that `pub` items with `#[no_mangle]` are exported from the dynamic library, and correctly excludes the items missing either of those annotations:
+
+```
+$ ONLY_DYNAMIC_SYMBOLS=1 ./script.sh | grep rfc3556
+=== DYNAMIC SYMBOLS ===
+0000000000054008 D foo_rfc3556_global_pub_with_no_mangle
+0000000000006f56 T foo_rfc3556_pub_with_no_mangle
+```
+
+However, there is a caveat - a collection of similar warnings from the compiler:
+
+```
+warning: `#[no_mangle]` has no effect on a foreign function
+  --> src/lib.rs:21:5
+   |
+21 |     #[no_mangle]
+   |     ^^^^^^^^^^^^ help: remove this attribute
+22 |     pub fn foo_rfc3556_pub_with_no_mangle();
+   |     ---------------------------------------- foreign function
+   |
+   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
+   = note: symbol names in extern blocks are not mangled
+```
+
+This warning is incorrect - `#[no_mangle]` does have an effect on a foreign function, and it's exactly what this RFC needs. The warning was added in a PR that [justifies the warning](https://github.com/rust-lang/rust/pull/86376#issuecomment-905864909) with the following reasoning:
+
+1. `#[no_mangle]` marks symbols as globally visible ("the behavior of `#[no_mangle]` causing symbols to be made global (regardless of rust's visibility rules) is not well explained")
+2. it does not make sense for imported symbols to be marked as globally visible ("It also does not make sense for those imports to additionally be exported by `#[no_mangle]`'s effects (and I would expect possible linker errors if they were)"
+
+Point 1 is reasonable - to some extent this RFC could be seen as partly rectifying that issue. Point 2, per the motivation section, isn't true in at least the linking scenario considered by this RFC.
+
+The implementation strategy for this RFC *on Linux* is:
+
+1. remove the above warning
+2. document the new behavior
+3. add a warning for re-exporting a symbol when the symbol is not globally visible
+4. add a warning for re-exporting a symbol when the library is not explicitly marked as a staticlib
+
+On point 3 - the goal is to catch cases where a) the user has specified non-staticlib link kind or b) the user has not specified any link kind. The semantics of case (a) are not defined in this RFC (see [future-possibilities](#future-possibilities)), and catching (b) is to avoid the user accidentally falling into case (a) where the linker selects a dylib automatically.
```

Comment:
Yes, I see this as a best-effort warning - if the `extern` block has no `link` attribute at all then there's not much we can do as that is a valid setup. But if you do have a link attribute on an extern block and that library is not explicitly static then possibly there's something wrong (there could also not be, e.g. if the function is actually from another lib rather than the one annotated on the extern block, but that seems unusual?)